core/trino-main/src/main/java/io/trino/execution/executor2/scheduler/SchedulingQueue.java

```diff
@@ -0,0 +1,531 @@
+ * weight in the long run is equal among groups. Within a group, the goal is to
+ * balance the consideration among tasks to ensure the accumulated weight in the
+ * long run is equal among tasks within the group.
+ *
+ * <p>Groups start in the blocked state and transition to the runnable state when a task is
+ * added via the {@link #enqueue(Object, Object, long)} method.</p>
+ *
+ * <p>Tasks are dequeued via the {@link #dequeue(long)}. When all tasks in a group have
+ * been dequeued, the group transitions to the running state and is removed from the
+ * queue.</p>
+ *
+ * <p>When a task time slice completes, it needs to be re-enqueued via the
+ * {@link #enqueue(Object, Object, long)}, which includes the desired
+ * increment in relative weight to apply to the task for further prioritization.
+ * The weight increment is also applied to the group.
+ * </p>
+ *
+ * <p>If a task blocks, the caller must call the {@link #block(Object, Object, long)}
+ * method to indicate that the task is no longer running. A weight increment can be
+ * included for the portion of time the task was not blocked.</p>
+ * <br/>
+ * <h2>Group state transitions</h2>
+ * <pre>
+ *                                                                 blockTask()
+ *    finishTask()               enqueueTask()                     enqueueTask()
+ *        ┌───┐   ┌──────────────────────────────────────────┐       ┌────┐
+ *        │   │   │                                          │       │    │
+ *        │   ▼   │                                          ▼       ▼    │
+ *      ┌─┴───────┴─┐   all blocked        finishTask()   ┌────────────┐  │
+ *      │           │◄──────────────O◄────────────────────┤            ├──┘
+ * ────►│  BLOCKED  │               │                     │  RUNNABLE  │
+ *      │           │               │   ┌────────────────►│            │◄───┐
+ *      └───────────┘       not all │   │  enqueueTask()  └──────┬─────┘    │
+ *            ▲             blocked │   │                        │          │
+ *            │                     │   │           dequeueTask()│          │
+ *            │ all blocked         ▼   │                        │          │
+ *            │                   ┌─────┴─────┐                  ▼          │
+ *            │                   │           │◄─────────────────O──────────┘
+ *            O◄──────────────────┤  RUNNING  │      queue empty     queue
+ *            │      blockTask()  │           ├───┐                 not empty
+ *            │                   └───────────┘   │
+ *            │                     ▲      ▲      │ finishTask()
+ *            └─────────────────────┘      └──────┘
+ *                not all blocked
+ * </pre>
+ *
+ * <h2>Implementation notes</h2>
+ * <ul>
+ *     <li>TODO: Initial weight upon registration</li>
+ *     <li>TODO: Weight adjustment during blocking / unblocking</li>
+ *     <li>TODO: Uncommitted weight on dequeue</li>
+ * </ul>
+ * </p>
+ */
+@NotThreadSafe
```

Comment:
Can you add this to PriorityQueue, and the other datastructures that are not threadsafe